README.md
# Nushell Plugin Python

[![PyPI version](https://badge.fury.io/py/nushell.svg)](https://pypi.org/project/nushell/)

This is a python module to allow for easy creation of plugins. It's
already fairly easy, but this can improve upon that! Each of the examples provided comes with:

 - the full example in a python file (without the .py extension)
 - containers with nushell to run with regular Python or (for some) a standalone binary (via pyinstaller)
 - README.md files to walk through the usage

Quick start examples for a filter and sink plugin are included below, and you
should see the [examples](examples) folder for the complete code, makefiles, and Dockerfiles.
Please contribute new examples along with tweaks to the code! I developed this in
under a day, so we can likely make it much better, and will need to as nushell
is developed.

## Install

You can install from pip

```bash
$ pip install nushell
```

or from source here:

```bash
$ git clone https://github.com/vsoch/nushell-plugin-python
cd nushell-plugin-python
python setup.py install
```

## Plugin Class Arguments

The following are shared parameters for both filter and sink plugins.

| Name | Description | Required or Default |
|------|-------------|-------------------|
| name | the name of the plugin | required |
| usage | plugin usage | required | 
| logging | enable logging to `nu_plugin_<name>.log` | defaults to True (enabled) |
| add_help | add the `--help` flag | defaults to True |
| parse_params | extract parameters from items | defaults to True |


## Parameters

Parameters can be named or positional arguments that are defined for your
plugin. You can actually have positional and named arguments with the
same name, and define the help string for one (and it will be used by
the other).

### Named Parameters

For either a sink or a filter, you can pass in named arguments, meaning that your
plugin will parse them like:

```
$ <plugin> --<name> <value>
```

or for a boolean (called a Switch)

```
$ <plugin> --<switch>
```

Here are examples of adding all kinds of named arguments

```python
# add_named_argument(name, argType, syntaxShape=None, usage=None)
plugin.add_named_argument("catch", "Switch", usage="catch a random pokemon")
plugin.add_named_argument("list", "Switch", usage="list pokemon names")
plugin.add_named_argument("list-sorted", "Switch", usage="list sorted names")
plugin.add_named_argument("avatar", "Optional", "String", "generate avatar")
plugin.add_named_argument("pokemon", "Optional", "String", "get pokemon")
```

For an argument to be Mandatory, just change "Optional" to "Mandatory" (the 
above plugin doesn't require anything).


### Positional Arguments

Positional arguments come after the plugin name, but don't have a flag.
For example:

```bash
$ <plugin> <positional>
```

Here are examples of adding positional arguments. Each needs a name,
if it's Mandatory or Optional (there is no Switch), the SyntaxShape (e.g. String, Int, Any)
and then a usage string.

```python
# add_positional_argument(name, argType, syntaxShape=None, usage=None)
plugin.add_positional_argument("number", "Mandatory", "Int", usage="number to parse")
plugin.add_positional_argument("secondNumber", "Optional", "Any", usage="second number to parse")
```

## Filter Plugin

A basic filter plugin will instantiate the `FilterPlugin` class, and then
provide a function to run for the filter.

```python
#!/usr/bin/env python3

from nushell.filter import FilterPlugin

# Your filter function will be called by the FilterPlugin, and should
# accept the plugin and the dictionary of params
def runFilter(plugin, params):
    '''runFilter will be executed by the calling SinkPlugin when method is "sink"
    '''
    # Get the string primitive passed by the user
    value = plugin.get_string_primitive()

    # Calculate the length
    intLength = len(value)

    # Print an integer response (can also be print_string_response)
    plugin.print_int_response(intLength)


# The main function is where you create your plugin and run it.
def main():

    # Initialize a new plugin
    plugin = FilterPlugin(name="len", 
                          usage="Return the length of a string")


    # Run the plugin by passing your filter function
    plugin.run(runFilter)


if __name__ == '__main__':
    main()
```

Notably, your filter function should taken a plugin and parsed command line
parameters (dictionary) as arguments. You can use the plugin to perform
several needed functions to send responses back to nushell, or log to `/tmp/nushell-plugin-<name>.log`:

```python
plugin.logger.<level>
plugin.get_string_primitive()
plugin.get_int_primitive()
plugin.print_int_response()
plugin.print_string_response()
```

### Examples

 - [len](examples/len) is a basic function to return the length of a string
 - [plus](examples/plus) adds two ints, and is an example with positional arguments


## Sink Plugin

A sink plugin will instantiate the `SinkPlugin` class, and then hand off
stdin (via a temporary file) to a sink function that you write.
Here is a dummy example.

```python
#!/usr/bin/env python3

from nushell.sink import SinkPlugin

# Your sink function will be called by the sink Plugin, and should
# accept the plugin and the dictionary of params
def sink(plugin, params):
    '''sink will be executed by the calling SinkPlugin when method is "sink"
    '''
    message = "Hello"
    excited = params.get("excited", False)
    name = params.get("name", "")
    
    # If we have a name, add to message
    message = "%s %s" %(message, name)
    
    # Are we excited?
    if excited:
        message += "!"

    print(message)


# The main function is where you create your plugin and run it.
def main():

    # Initialize a new plugin
    plugin = SinkPlugin(name="hello", 
                        usage="A friendly plugin")


    # Add named arguments (notice we check for in params in sink function)
    # add_named_argument(name, argType, syntaxShape=None, usage=None)
    plugin.add_named_argument("excited", "Switch", usage="add an exclamation point!")
    plugin.add_named_argument("name", "Optional", "String", usage="say hello to...")

    # Run the plugin by passing your sink function
    plugin.run(sink)


if __name__ == '__main__':
    main()
```

### Parameters

Since you can pipe content into a sink, the piped content is parsed into a list
and passed with params as the `_pipe` key. For example, if we do:

```bash
> ls | get name | hello --name Dinosaur
Hello Dinosaur
```

And then look in the output file, we see that the parsed params include a pipe
of all the named of the listed files (that we generated above)

```bash
PARAMS {'name': 'Dinosaur', '_pipe': ['Makefile', 'README.md', 'Dockerfile', 'nu_plugin_hello', 'Dockerfile.standalone']}
```

If you don't want to parse the pipe, set your plugin.parse_pipe to False:

```python
plugin.parse_pipe = False
plugin.run(sink)
```

And the result will include the full list of entries with items and tags.

```python
PARAMS {'name': 'Dinosaur', '_pipe': [[{'tag': {'anchor': None, 'span': {'start': 0, 'end': 2}}, 'item': {'Primitive': {'String': 'Makefile'}}}, {'tag': {'anchor': None, 'span': {'start': 0, 'end': 2}}, 'item': {'Primitive': {'String': 'README.md'}}}, {'tag': {'anchor': None, 'span': {'start': 0, 'end': 2}}, 'item': {'Primitive': {'String': 'Dockerfile'}}}, {'tag': {'anchor': None, 'span': {'start': 0, 'end': 2}}, 'item': {'Primitive': {'String': 'nu_plugin_hello'}}}, {'tag': {'anchor': None, 'span': {'start': 0, 'end': 2}}, 'item': {'Primitive': {'String': 'Dockerfile.standalone'}}}]]}
```

### Examples

 - [pokemon](examples/pokemon) ascii pokemon on demand!
 - [hello](examples/hello) say hello using a sink!


## Single Binary

In that you are able to compile your module with [pyinstaller](https://pyinstaller.readthedocs.io/en/stable/operating-mode.html) (e.g., see [examples/len](examples/len)) you can build your python script as a simple binary, and one that doesn't even need nushell installed as a module anymore. Why might you want to do this? It will mean that your plugin is a single file (binary) and you don't need to rely on modules elsewhere in the system. I suspect there are other ways to compile
python into a single binary (e.g., cython) but this was the first I tried, and fairly straight forward.
If you find a different or better way, please contribute to this code base!

**Important** I've found that modules with added data files don't do well (an example is [pokemon](examples/pokemon) here!)
and that's why we can't uninstall pokemon or nushell. However, I think you would have luck with most text based,
simple modules. And of course, you don't have to do this! It's totally ok to keep your Python modules
installed alongside nushell, and used when your plugin is executed.

## License

This code is licensed under the MPL 2.0 [LICENSE](LICENSE).

## Help and Contribution

Please contribute to the package, or post feedback and questions as <a href="https://github.com/vsoch/nushell-plugin-python/issues" target="_blank">issues</a>.


CHANGELOG.md
# CHANGELOG

This is a manually generated log to track changes to the repository for each release. 
Each section should include general headers such as **Implemented enhancements** 
and **Merged pull requests**. All closed issued and bug fixes should be 
represented by the pull requests that fixed them.
Critical items to know are:

 - renamed commands
 - deprecated / removed commands
 - changed defaults
 - backward incompatible changes
 - migration guidance
 - changed behaviour 

Versions here coincide with releases on pypi.

## [master](https://github.com/vsoch/nushell-plugin-python)
 - basic testing (0.0.16)
 - allowing end-filter to return help response (0.0.15)
 - adding general function to get any primitive (0.0.14)
 - parsing of positional arguments (0.0.13)
 - support for sink pipes (0.0.12)
 - adding standalone build examples (0.0.11)
 - adding base of package and examples (0.0.1)
 - skeleton of package while server under development (0.0.0)


nushell/tests/helpers.py
#!/usr/bin/python

# Copyright (C) 2019-2020 Vanessa Sochat.

# This Source Code Form is subject to the terms of the
# Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
# with this file, You can obtain one at http://mozilla.org/MPL/2.0/.

def assert_good_response(response):
    '''ensure that a response is good, meaning jsonrpc 2.0 and method response
    '''
    for key in ['jsonrpc', 'method', 'params']:
        assert key in response

    assert isinstance(response['params'], (dict, list))
    assert response['method'] == 'response'


def check_plugin_config(plugin, plugin_name, usage, is_filter):
    '''a helper function to test a general plugin configuration
    '''
 
    # Before adding args or getting help, we don't have any parameters
    assert not plugin.argUsage
    assert not plugin.positional
    assert not plugin._positional
    assert not plugin.named

    # Help shouldn't be in current help
    assert "--help" not in plugin.get_help()

    # Once we generate a config (like the first call) we should have help
    plugin_config = plugin.get_config()
    assert "help" in plugin.named

    for key in ['name', 'usage', 'positional', 'rest_positional', 'named', 'is_filter']:
        assert key in plugin_config

    assert plugin_config["is_filter"] == is_filter
    assert plugin_config["usage"] == usage
    assert plugin_config["name"] == plugin_name
    assert plugin.name == plugin_name
    assert plugin.usage == usage

    # The name and usage should be included, and now --help is added too
    for contender in [usage, plugin_name, '--help', 'show this usage']:
        assert contender in plugin.get_help()

def check_remove_help(plugin, plugin_name, usage, is_filter):
    '''similar checks, but remove the help parameters
    '''
    # But if we requested to not add help, shouldn't be there
    plugin_config = plugin.get_config()
    assert "help" not in plugin.named
    assert "--help" not in plugin.get_help()

    # Now add named arguments
    plugin.add_named_argument("greet", "Switch", usage="say hello")
    assert "greet" in plugin.argUsage
    assert "greet" in plugin.named

    # The user can optionally just give a name (a positional argument)
    plugin.add_positional_argument("avatar", "Optional", "String")
    assert "avatar" in plugin._positional
    assert plugin.positional # len > 0


nushell/tests/test_sink.py
#!/usr/bin/python

# Copyright (C) 2019-2020 Vanessa Sochat.

# This Source Code Form is subject to the terms of the
# Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
# with this file, You can obtain one at http://mozilla.org/MPL/2.0/.

from nushell.sink import SinkPlugin
from .helpers import (
    assert_good_response,
    check_plugin_config,
    check_remove_help
)
from .plugin_requests import (
    config_request,
    sink_named_request,
    sink_help_request
)

import os
import pytest


def sink(plugin, params):
    '''sink will be executed by the calling SinkPlugin when method is "sink"
       instead of just printing, we return params to test
    '''
    return params


def test_sink_name(tmp_path):
    '''ensure that a plugin's name is made all lowercase, with spaces removed
       note that the user is allowed to use special characters.
    '''
    # Name must be made all lowercase
    plugin_name = "sink"
    usage = "A dummy sink plugin to print to the terminal"
    plugin = SinkPlugin(name=plugin_name, usage=usage, logging=False)

    assert plugin._clean_name("a a a") == "a-a-a"
    assert plugin._clean_name("SINK") == "sink"

def test_sink_plugin(tmp_path):
    '''test creation of a simple sink plugin
    '''
    plugin_name = "sink"
    usage = "A dummy sink plugin to print to the terminal"
    plugin = SinkPlugin(name=plugin_name, usage=usage, logging=False)

    # Test request for config
    response = plugin.test(sink, config_request)
    assert_good_response(response)
    assert plugin.get_config() == response['params']['Ok']

    # Test help request (should return help)
    response = plugin.test(sink, sink_help_request)
    assert plugin.name in response and plugin.usage in response
 
    # Test "sink" named arg request (should return params from sink())
    # {'sink': True, '_positional': [], '_pipe': []}
    response = plugin.test(sink, sink_named_request)
    for key in ['sink', '_positional', "_pipe"]:
        assert key in response

    
def test_sink_config(tmp_path):
    '''test configure of a simple sink plugin
    '''
    plugin_name = "sink"
    usage = "A dummy sink plugin to print to the terminal"
    plugin = SinkPlugin(name=plugin_name, usage=usage, logging=False)
    check_plugin_config(plugin, plugin_name, usage, is_filter=False)

    # Test without adding help
    plugin = SinkPlugin(name=plugin_name, usage=usage, logging=False, add_help=False)
    check_remove_help(plugin, plugin_name, usage, is_filter=False)


nushell/tests/test_filter.py
#!/usr/bin/python

# Copyright (C) 2019-2020 Vanessa Sochat.

# This Source Code Form is subject to the terms of the
# Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
# with this file, You can obtain one at http://mozilla.org/MPL/2.0/.

from nushell.filter import FilterPlugin
from .helpers import (
    assert_good_response,
    check_plugin_config,
    check_remove_help
)
from .plugin_requests import (
    config_request,
    filter_begin_request,
    filter_end_request,
    filter_string_request,
    filter_int_request,
    filter_custom_request
)

import os
import pytest


def func(plugin, params):
    '''func will be executed by the calling FilterPlugin when method is "filter"
       instead of printing responses, we return them to the caller for testing
    '''
    return params


def test_filter_plugin(tmp_path):
    '''test creation of a simple sink plugin
    '''
    plugin_name = "filter"
    usage = "A dummy filter plugin to print to the terminal"
    plugin = FilterPlugin(name=plugin_name, usage=usage, logging=False)

    # Before config, we haven't cached any params
    for attr in ['params', 'args']:
        assert not getattr(plugin, attr, None)

    # Test request for config
    response = plugin.test(func, config_request)
    assert_good_response(response)
    assert plugin.get_config() == response['params']['Ok']

    # After config, we have empty params / args
    for attr in ['params', 'args']:
        assert getattr(plugin, attr) in [[], {}]

    # Check begin_filter
    response = plugin.test(func, filter_begin_request)
    assert_good_response(response)
    assert response['params'] == {'Ok': []}

    # Now params should be populated
    assert 'args' in plugin.params
    assert 'help' in plugin.args and '_positional' in plugin.args

    # Testing the filter method with string request
    response = plugin.test(func, filter_string_request)
    assert plugin.get_string_primitive() == 'pancakes'

    # Filters differ from sinks in tha they don't return _pipe
    for key in ['help', '_positional']:
        assert key in response

    # Testing the filter method with int request
    response = plugin.test(func, filter_int_request)
    assert plugin.get_int_primitive() == 'imanumber'

    # Testing the filter method with custom request
    response = plugin.test(func, filter_custom_request)
    assert plugin.get_primitive("Any") == 'imathing'

    # Test end_filter request     
    response = plugin.test(func, filter_end_request)
 


def test_filter_config(tmp_path):
    '''test configure of a simple sink plugin
    '''
    plugin_name = "filter"
    usage = "A dummy filter plugin to print to the terminal"
    plugin = FilterPlugin(name=plugin_name, usage=usage, logging=False)
    check_plugin_config(plugin, plugin_name, usage, is_filter=True)

    # Test without adding help
    plugin = FilterPlugin(name=plugin_name, usage=usage, logging=False, add_help=False)
    check_remove_help(plugin, plugin_name, usage, is_filter=True)


nushell/sink.py

# Copyright (C) 2019 Vanessa Sochat.

# This Source Code Form is subject to the terms of the
# Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
# with this file, You can obtain one at http://mozilla.org/MPL/2.0/.


from nushell.plugin import PluginBase

import fileinput
import json


class SinkPlugin(PluginBase):
    '''A sink plugin is identified by is_filter set to false in the
       configuration, and interacts with nushell by way of nushell
       asking for the configuration upon discovery on the path (method "config")
       and then passing a temporary file as argument (method "sink")
    '''
    is_filter = False
    parse_pipe = True

    def get_sink_params(self, input_params):
        '''The input params (under ["params"] is a list, with the first entry
           being the args dictionary (that we pass to self.parse_params) and
           the remaining being entries that are passed if the sink is used as 
           a pipe. If not, it looks like an empty list, like below:

          [{'args': {'positional': None,
	     'named': {'switch': {'tag': {'anchor': None,
	        'span': {'start': 58, 'end': 64}},
	       'item': {'Primitive': {'Boolean': True}}},
	      'mandatory': {'tag': {'anchor': None, 'span': {'start': 20, 'end': 32}},
	       'item': {'Primitive': {'String': 'MANDATORYARG'}}},
	      'optional': {'tag': {'anchor': None, 'span': {'start': 44, 'end': 55}},
	       'item': {'Primitive': {'String': 'OPTIONALARG'}}}}},
	    'name_tag': {'anchor': None, 'span': {'start': 0, 'end': 7}}},
	   []]
        '''
        if not input_params:
            return input_params 

        # Args are always the first entry
        args = input_params.pop(0)
        params = self.parse_params(args)

        # The pipe entries are the rest (pass as _pipe)
        params["_pipe"] = self._parse_pipe(input_params)
        return params


    def _parse_pipe(self, pipeList):
        '''parse the list of piped input, typically this means string that
           have come from the terminal. To disable this, set the client
           parse_pipe to False.

           Parameters
           ==========
           pipeList: is the second index of the "params" dict from the request
        '''
        # No pipe will produce empty list
        if not pipeList or not self.parse_pipe:
            return pipeList

        pipeList = pipeList.pop(0)
        return self.parse_primitives(pipeList)


    def test(self, sinkFunc, line):
        '''test is akin to run, but instead of printing a result for the user,
           we return to the calling function. A line to parse is also required.
           since it's coming from Python, we also assume that we don't need to
           json.loads() the line from a string (it's a dictionary)
        '''
        method = line.get("method")

        # Case 1: Nu is asking for the config to discover the plugin
        if method == "config":
            plugin_config = self.get_config()
            return self.get_good_response(plugin_config)
            
        # Case 3: A filter must return the item filtered with a tag
        elif method == "sink":

            # Parse parameters for the calling sink, _pipe included
            params = self.get_sink_params(line['params'])

            # The only case of not running is if the user asks for help
            if params.get('help', False):
                return self.get_help()

            # Run the sink, and provide the user with plugin and params
            return sinkFunc(self, params)


    def run(self, sinkFunc):
        '''the main run function is required to take a user sinkFunc.
        '''
        for line in fileinput.input():

            x = json.loads(line)
            method = x.get("method")

            # Keep log of requests from nu
            self.logger.info("REQUEST %s" % line)
            self.logger.info("METHOD %s" % method)

            # Case 1: Nu is asking for the config to discover the plugin
            if method == "config":
                plugin_config = self.get_config()
                self.logger.info("plugin-config: %s" % json.dumps(plugin_config))
                self.print_good_response(plugin_config)
                break

            # Case 2: A sink passes execution to the sink function
            elif method == "sink":

                # Parse parameters for the calling sink, _pipe included
                params = self.get_sink_params(x['params'])
                self.logger.info("PARAMS %s" % params)

                # The only case of not running is if the user asks for help
                if params.get('help', False):
                    print(self.get_help())

                # Run the sink, and provide the user with plugin and params
                else:
                    sinkFunc(self, params)
                break


nushell/plugin.py

# Copyright (C) 2019 Vanessa Sochat.

# This Source Code Form is subject to the terms of the
# Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
# with this file, You can obtain one at http://mozilla.org/MPL/2.0/.


from nushell.logger import NushellLogger

import json
import os
import sys
import tempfile


class PluginBase:
    '''a PluginBase includes a name, usage, and is the base class for both
       a sink and filter plugin
    '''
    def __init__(self, name, usage, 
                 logging=True, add_help=True, parse_params=True):

        '''Set the name and usage to generate the configuration

           Parameters
           ==========
           name: the name provided by the user
           usage: the plugin usage, should be one line
           logging: if True, will output logfile to /tmp/nu_plugin_<name>.log
           add_help: if True, adds a custom --help command (unless defined)
           parse_params: extract values from "args" (don't return raw)
        '''
        self.name = self._clean_name(name)
        self.usage = usage
        self.positional = []
        self._positional = [] # list of names
        self.rest_positional = None
        self.named = {}
        self.argUsage = {}
        self.logger = self.get_logger(logging)
        self.add_help = add_help
        self._parse_params = parse_params

# Arguments

    def _check_argument(self, name, argType, syntaxShape):
        '''check an argument (named or positional) to ensure that the name
           is lowercase without spaces, and the syntaxShape is valid
           return the cleaned values.
        '''
        name = self._clean_name(name)
        argType = self._camel_case(argType)

        # If syntaxShape provided, ensure is valid (warn user if not)
        if syntaxShape:
            syntaxShape = self._get_syntax_shape(syntaxShape)

        # Return the parse argument
        arg = {"name": name,
               "type": argType,
               "shape": syntaxShape}

        return arg


    def add_positional_argument(self, name, argType, syntaxShape=None, usage=None):
        '''a positional argument doesn't require a flag, but is provided
           based on a positional index
        '''
        # includes name, type, shape
        arg = self._check_argument(name, argType, syntaxShape)

        # Syntax shape must be provided for Optional/Mandatory
        if arg['type'] not in ["Optional", "Mandatory"]:
            self.logger.exit("positional type can only be Optional/Mandatory")

        # Mandatory can have no syntax shape, Any or Block
        if not arg['shape']:
            self.positional.append({arg["type"]: [arg["name"]]})
        else:
            self.positional.append({arg["type"]: [arg["name"], arg["shape"]]})

        # Add to list of names, we use this to add to --help
        self._positional.append(arg['name'])

        # Add the usage, if defined
        if usage is not None:
            self.argUsage[arg['name']] = usage

        self.logger.debug("Updated positional arguments %s" % self._positional)


    def add_named_argument(self, name, argType, syntaxShape=None, usage=None):
        '''adding a named argument comes down to one of:
           Switch: akin to a boolean, doesn't require a SyntaxShape
           Optional: an optional argument (requires Syntax Shape)
           Mandatory: required! (also requires a SyntaxShape)

           Parameters
           ==========
           name: the name of the argument, should be lowercase without spaces
           argType: the arg type, must be one of Switch, Optional, Mandatory 
           syntaxShape: if not Switch, must provide syntaxShape.
           usage: if provided, is added to argUsage for custom --help
        '''
        # Syntax shape must be provided for Optional/Mandatory
        if argType in ["Optional", "Mandatory"] and not syntaxShape:
            self.logger.exit("SyntaxShape is required for Optional/Mandatory")

        # includes name, type, shape
        arg = self._check_argument(name, argType, syntaxShape)

        if arg['type'] == "Switch":
            self.named[arg['name']] = "Switch"
        elif arg['type'] in ["Optional", "Mandatory"]:
            self.named[arg['name']] = {arg['type']: arg['shape']}

        # Add usage, if provided
        if usage:
            self.argUsage[arg['name']] = usage

        self.logger.debug("Updated named arguments %s" % self.named)


    def _get_syntax_shape(self, shape):
        '''return a valid syntax shape, warn the user if the provided type 
           is not found at see src/parser/hir/syntax_shape.rs
           at https://github.com/nushell/nushell
        '''
        shape = self._camel_case(shape)
        shapes = ["Any", "List", "String", "Member", "ColumnPath", 
                  "Number", "Int", "Path", "Pattern", "Block"]

        if shape not in shapes:
            self.logger.warning("%s is not a valid SyntaxShape" % shape)
            return "String"
        return shape


# Print / Logging

    def get_logger(self, logging):
        '''if logging is not enabled, set to quiet.
        '''
        logname = "nu_plugin_%s.log" % self.name
        logfile = os.path.join(tempfile.gettempdir(), logname)
        if not logging:
            return NushellLogger(logfile, level=0)
        return NushellLogger(logfile)


    def get_good_response(self, response):
        '''generate a good response. Confirming to jsonprc 2.0, we include a 
           method "response" and params that should be a dict with key "Ok" and 
           value as a config, an empty list (begin or end filter) or a list
           of dict responses (for a filter)
        '''
        json_response = {
            "jsonrpc": "2.0",
            "method": "response",
            "params": {"Ok": response}
        }
        return json_response


    def print_good_response(self, response):
        '''generate and print a good response.
        '''
        json_response = self.get_good_response(response)
        self.logger.info("Printing response %s" % response)
        print(json.dumps(json_response))
        sys.stdout.flush()


# Configuration

 
    def get_config(self):
        '''return configuration object, is_filter must be defined by subclass
           note that get_config is the first call to any kind of plugin,
           so here is where we add a help argument if it's not added.
        '''
        # If help not in named, add it.
        if "help" not in self.named and self.add_help:
            self.named['help'] = "Switch"
            self.argUsage['help'] = "show this usage"

        return {
            "name": self.name,
            "usage": self.usage,
            "positional": self.positional,
            "rest_positional": self.rest_positional,
            "named": self.named,
            "is_filter": self.is_filter}


# Parsing, Help and Tags

    def parse_primitives(self, listing):
        '''given a listing of primitives (e.g., a pipelist from _parse_pipe
           or positional arguments from parse_params) return the content
           of the primitive, regardless of type. Input should look like:
   
            [{"tag":
               {"anchor":null,"span":{"start":5,"end":6}},
               "item":{"Primitive":{"Int":1}}}..]
        '''
        # In case None
        listing = listing or []

        # Return list of values as the pipe content
        entries = []

        # Each entry has a tag and item. We want the Primitive (type)
        for entry in listing:
            item = entry['item'].get('Primitive')
            entries = entries + list(item.values())

        return entries


    def getTag(self):
        '''for local testing without Nu, we provide a function to return
           a dummy tag
        '''
        return {"anchor":None, "span":{"end":0, "start":0}}


    def parse_params(self, input_params):
        '''parse the parameters into an easier to parse object. An example looks 
           like the following. For a sink - this is the first item in a list
           under params. For a filter, it's a dict directly under params.

          {'args': {'positional': None,
	     'named': {'switch': {'tag': {'anchor': None,
	        'span': {'start': 58, 'end': 64}},
	       'item': {'Primitive': {'Boolean': True}}},
	      'mandatory': {'tag': {'anchor': None, 'span': {'start': 20, 'end': 32}},
	       'item': {'Primitive': {'String': 'MANDATORYARG'}}},
	      'optional': {'tag': {'anchor': None, 'span': {'start': 44, 'end': 55}},
	       'item': {'Primitive': {'String': 'OPTIONALARG'}}}}},
	    'name_tag': {'anchor': None, 'span': {'start': 0, 'end': 7}}},
        '''
        if not input_params:
            return input_params 

        if "args" not in input_params:
            return input_params

        # Does the plugin want to skip parsing params?
        if not self._parse_params:
            return input_params

        positional = input_params['args'].get('positional', [])
        named = input_params['args'].get('named', {})

        # We will return lookup dictionary of params
        params = {}

        # Keep a simple dictionary with values we know types for
        for name, values in named.items():

            # is it a String? Boolean?
            value_type = list(values['item']['Primitive'].keys())[0]

            if value_type == "String":
                params[name] = values['item']['Primitive']['String']

            elif value_type == "Boolean":
                params[name] = values['item']['Primitive']['Boolean']

            # If you use other types, add them here
            else:
                self.logger.info("Invalid paramater type %s:%s" %(name, values))

        # Add positional arguments
        params["_positional"] = self.parse_primitives(positional)
        return params        


    def get_help(self):
        '''based on the named and positional arguments, we provide a usage
           string that can be used for a custom <name> --help function.
        '''
        args = ""

        # Positional arguments
        if self._positional:
            for name in self._positional:

                argString = "%s %s" %(name, name.upper())

                spaces = 19 - len(argString)
                args = args + argString + " "*spaces

                # If a usage is defined
                if self.argUsage.get(name) is not None:
                    args += " %s" % self.argUsage.get(name)
                args += "\n"                    

        # Named arguments
        for name, argType in self.named.items():

            # If it's a mandatory or optional
            if isinstance(argType, dict):
                argString = "--%s %s" %(name, name.upper())
            else:
                argString = "--%s" % name         

            # Ensure nice spacing (might need to adjust)
            spaces = 19 - len(argString)
            args = args + argString + " "*spaces
            
            # If a usage is defined
            if self.argUsage.get(name) is not None:
                args += " %s" % self.argUsage.get(name)
            args += "\n"                    

        return "%s: %s\n\n%s\n" %(self.name, self.usage, args)


# Helpers

    def _clean_name(self, name):
        '''the name must be all lowercase, and with no spaces.

           Parameters
           ==========
           name: the name provided by the user
        '''
        return name.replace(' ', '-').lower()


    def _camel_case(self, name):
        '''nushell uses camel case (or all lowercase with capital letters)
        '''
        return name.lower().capitalize()
